File: src/test/scala/DeflaterInflaterStreamTest.scala
/* -*-mode:scala; c-basic-offset:2; indent-tabs-mode:nil -*- */
package com.jcraft.jsch.jzlib

import org.scalatest._
import org.scalatest.flatspec._
import matchers.should._

import scala.language.reflectiveCalls

import java.io.{ByteArrayOutputStream => BAOS, ByteArrayInputStream => BAIS}

class DeflaterInflaterStreamTest extends AnyFlatSpec with BeforeAndAfter with Matchers {

  before {
  }

  after {
  }

  behavior of "Deflter and Inflater"

  it can "deflate and infate data one by one." in {
    val data1 = randombuf(1024)
    implicit val buf = new Array[Byte](1)

    val baos = new BAOS
    val gos = new DeflaterOutputStream(baos)
    data1 -> gos
    gos.close

    val baos2 = new BAOS
    new InflaterInputStream(new BAIS(baos.toByteArray)) -> baos2
    val data2 = baos2.toByteArray 

    data2.length should equal (data1.length)
    data2 should equal (data1)
  }

  behavior of "DeflterOutputStream and InflaterInputStream"

  it can "deflate and infate." in {

    (1 to 100 by 3).foreach { i =>

      implicit val buf = new Array[Byte](i)

      val data1 = randombuf(10240)

      val baos = new BAOS
      val gos = new DeflaterOutputStream(baos)
      data1 -> gos
      gos.close

      val baos2 = new BAOS
      new InflaterInputStream(new BAIS(baos.toByteArray)) -> baos2
      val data2 = baos2.toByteArray

      data2.length should equal (data1.length)
      data2 should equal (data1)
    }
  }

  behavior of "Deflter and Inflater"

  it can "deflate and infate nowrap data." in {

    (1 to 100 by 3).foreach { i =>

      implicit val buf = new Array[Byte](i)

      val data1 = randombuf(10240)

      val baos = new BAOS
      val deflater = new Deflater(JZlib.Z_DEFAULT_COMPRESSION,
                                 JZlib.DEF_WBITS,
                                 true)
      val gos = new DeflaterOutputStream(baos, deflater)
      data1 -> gos
      gos.close

      val baos2 = new BAOS
      val inflater = new Inflater(JZlib.DEF_WBITS, true)
      new InflaterInputStream(new BAIS(baos.toByteArray), inflater) -> baos2
      val data2 = baos2.toByteArray

      data2.length should equal (data1.length)
      data2 should equal (data1)
    }
  }

  it can "deflate and infate nowrap data with MAX_WBITS." in {
    implicit val buf = new Array[Byte](100)

    List(randombuf(10240),
         """{"color":2,"id":"EvLd4UG.CXjnk35o1e8LrYYQfHu0h.d*SqVJPoqmzXM::Ly::Snaps::Store::Commit"}""".getBytes) foreach { data1 =>

      val deflater = new Deflater(JZlib.Z_DEFAULT_COMPRESSION,
                                  JZlib.MAX_WBITS,
                                  true)

      val inflater = new Inflater(JZlib.MAX_WBITS, true)

      val baos = new BAOS
      val gos = new DeflaterOutputStream(baos, deflater)
      data1 -> gos
      gos.close

      val baos2 = new BAOS
      new InflaterInputStream(new BAIS(baos.toByteArray), inflater) -> baos2
      val data2 = baos2.toByteArray

      data2.length should equal (data1.length)
      data2 should equal (data1)
    }
  }
}

File: src/test/java/com/jcraft/jsch/jzlib/DeflaterInflaterStreamTest.java
package com.jcraft.jsch.jzlib;

import static com.jcraft.jsch.jzlib.Package.*;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Arrays;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class DeflaterInflaterStreamTest {

  @BeforeEach
  public void before() {}

  @AfterEach
  public void after() {}

  @Test
  public void testDeflaterAndInflaterCanDeflateAndInflateDataOneByOne() throws IOException {
    byte[] data1 = randombuf(1024);
    byte[] buf = new byte[1];

    ByteArrayOutputStream baos = new ByteArrayOutputStream();
    DeflaterOutputStream gos = new DeflaterOutputStream(baos);
    readArray(data1, gos, buf);
    gos.close();

    ByteArrayOutputStream baos2 = new ByteArrayOutputStream();
    readIS(new InflaterInputStream(new ByteArrayInputStream(baos.toByteArray())), baos2, buf);
    byte[] data2 = baos2.toByteArray();

    assertEquals(data1.length, data2.length);
    assertArrayEquals(data1, data2);
  }

  @Test
  public void testDeflaterOutputStreamAndInflaterInputStreamCanDeflateAndInflate()
      throws IOException {

    for (int i = 1; i < 100; i += 3) {

      byte[] buf = new byte[i];

      byte[] data1 = randombuf(10240);

      ByteArrayOutputStream baos = new ByteArrayOutputStream();
      DeflaterOutputStream gos = new DeflaterOutputStream(baos);
      readArray(data1, gos, buf);
      gos.close();

      ByteArrayOutputStream baos2 = new ByteArrayOutputStream();
      readIS(new InflaterInputStream(new ByteArrayInputStream(baos.toByteArray())), baos2, buf);
      byte[] data2 = baos2.toByteArray();

      assertEquals(data1.length, data2.length);
      assertArrayEquals(data1, data2);
    }
  }

  @Test
  public void testDeflaterAndInflaterCanDeflateAndInflateNowrapData() throws IOException {

    for (int i = 1; i < 100; i += 3) {

      byte[] buf = new byte[i];

      byte[] data1 = randombuf(10240);

      ByteArrayOutputStream baos = new ByteArrayOutputStream();
      Deflater deflater = new Deflater(JZlib.Z_DEFAULT_COMPRESSION, JZlib.DEF_WBITS, true);
      DeflaterOutputStream gos = new DeflaterOutputStream(baos, deflater);
      readArray(data1, gos, buf);
      gos.close();

      ByteArrayOutputStream baos2 = new ByteArrayOutputStream();
      Inflater inflater = new Inflater(JZlib.DEF_WBITS, true);
      readIS(new InflaterInputStream(new ByteArrayInputStream(baos.toByteArray()), inflater), baos2,
          buf);
      byte[] data2 = baos2.toByteArray();

      assertEquals(data1.length, data2.length);
      assertArrayEquals(data1, data2);
    }
  }

  @Test
  public void testDeflaterAndInflaterCanDeflateAndInflateNowrapDataWithMaxWbits() {
    byte[] buf = new byte[100];

    Arrays.asList(randombuf(10240),
        "{\"color\":2,\"id\":\"EvLd4UG.CXjnk35o1e8LrYYQfHu0h.d*SqVJPoqmzXM::Ly::Snaps::Store::Commit\"}"
            .getBytes())
        .forEach(uncheckedConsumer(data1 -> {
          Deflater deflater = new Deflater(JZlib.Z_DEFAULT_COMPRESSION, JZlib.MAX_WBITS, true);

          Inflater inflater = new Inflater(JZlib.MAX_WBITS, true);

          ByteArrayOutputStream baos = new ByteArrayOutputStream();
          DeflaterOutputStream gos = new DeflaterOutputStream(baos, deflater);
          readArray(data1, gos, buf);
          gos.close();

          ByteArrayOutputStream baos2 = new ByteArrayOutputStream();
          readIS(new InflaterInputStream(new ByteArrayInputStream(baos.toByteArray()), inflater),
              baos2, buf);
          byte[] data2 = baos2.toByteArray();

          assertEquals(data1.length, data2.length);
          assertArrayEquals(data1, data2);
        }));
  }
}
